gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/tasks/ResolveProductDependenciesTask.java

```diff
@@ -0,0 +1,350 @@
+                        "Encountered product dependency declaration that was also declared as optional for '%s', "
+                                + "either remove the dependency or optional declaration",
+                        productId));
+            }
+            allProductDependencies.merge(productId, declaredDep, ProductDependencyMerger::merge);
+            if (declaredDep.getOptional()) {
+                log.trace("Product dependency for '{}' declared as optional", productId);
+                allOptionalDependencies.add(productId);
+            }
+        });
+
+        // Merge all discovered and declared product dependencies
+        dedupDiscoveredProductDependencies().forEach(discoveredDependency -> {
+            ProductId productId = ProductId.of(discoveredDependency);
+            if (getIgnoredProductIds().get().contains(productId)) {
+                log.trace("Ignored product dependency for '{}'", productId);
+                return;
+            }
+            allProductDependencies.merge(productId, discoveredDependency, (declaredDependency, _newDependency) -> {
+                ProductDependency mergedDependency =
+                        ProductDependencyMerger.merge(declaredDependency, discoveredDependency);
+                if (mergedDependency.equals(discoveredDependency)) {
+                    getLogger()
+                            .error(
+                                    "Please remove your declared product dependency on '{}' because it is"
+                                            + " already provided by a jar dependency:\n\n"
+                                            + "\tProvided:     {}\n"
+                                            + "\tYou declared: {}",
+                                    productId,
+                                    discoveredDependency,
+                                    declaredDependency);
+                }
+                return mergedDependency;
+            });
+        });
+
+        // Ensure optional product dependencies are marked as such.
+        allOptionalDependencies.forEach(
+                productId -> allProductDependencies.computeIfPresent(productId, (_productId, existingDep) -> {
+                    log.trace("Product dependency for '{}' set as optional", productId);
+                    existingDep.setOptional(true);
+                    return existingDep;
+                }));
+
+        List<ProductDependency> productDeps = allProductDependencies.values().stream()
+                .sorted(Comparator.comparing(ProductDependency::getProductGroup)
+                        .thenComparing(ProductDependency::getProductName))
+                .collect(ImmutableList.toImmutableList());
+
+        Serializations.writeProductDependencyReport(
+                ProductDependencyReport.of(productDeps),
+                getOutputFile().getAsFile().get());
+    }
+
+    private Set<ProductDependency> dedupDiscoveredProductDependencies() {
```

Comment:
I think we want all of this logic to occur inside of CreateManifestTask so that regardless of how product dependencies are received we don't add self-dependencies or duplicate ones

Reply:
that really feels like we're splitting the logic and constraints for the same unit of work into two places. Why pull stuff out to another task if we're going to still have to have code in CreateManifest to double check things?

Reply:
we pulled it out so that it was easy to replace/provide an additional implementation. It seems reasonable for createManifest to ensure that the preconditions it requires for writing to manifest/lock file are met

Reply:
Otherwise you'd have to implement the same deduplication/filtering logic in each seperate mechanism for discovering dependencies

Reply:
the way the resolve task is written, if someone uses it and overrides the discovered dependencies, it will still provide errors and warnings about duplicates and self-dependencies. So we have the same protections (and useful messages) as now. 

I could see the createManifest task doing a final validation on the provided list and throwing errors if there is a self-reference or a duplicate. That way if someone workflow doesn't use the resolve task at all, the system will fail fast and force them to deconflict things with conscious decisions rather than by hiding potential errors. 

Good?

Reply:
ya thats great thanks!